File: nuido_flow_network/static/src/models/network/http_header_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { NodeModel } from "@nuido/models/node";
import { HttpHeaderPort } from "@nuido_flow_network/components/ports/http_header_port";
export class HttpHeaderNodeModel extends NodeModel {
    setup() {
        const auxOutId = "aux-out-" + this.id + "-1";
        this.addAuxOutPort(auxOutId, HttpHeaderPort.name, 1, {
            role: "http-header"
        });
        this.http_headers = [];
    }
    addOrUpdateHeader(name, value, is_mask_value) {
        const field = this.http_headers.find(o => o.name === name);
        if (field) {
            field.value = value;
        }
        else {
            this.http_headers.push({
                name: name,
                value: value,
                is_mask_value: is_mask_value
            });
        }
    }
    removeHeader(name) {
        const field_idx = this.http_headers.findIndex(o => o.name === name);
        if (field_idx > -1) {
            this.http_headers.splice(field_idx, 1);
        }
    }
}

File: nuido_flow_network/flows/network/http_header_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.
import logging

from odoo.tools import safe_eval

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode
from odoo.addons.nuido_flow.flows.tools.tools import get_default_context_for_eval, get_active_record_info

from odoo.addons.nuido_flow.flows.tools.log_const import LOGGER_NAME
_logger = logging.getLogger(LOGGER_NAME)

class HttpHeaderNode(BaseNode):
    def _process(self, params) -> any:
        super()._process(params)

        context = get_default_context_for_eval(self.env)
        if params is not None:
            context['params'] = params

        info = get_active_record_info(self.env)
        context = {**context, **info}
        res = {}

        try:
            for h in self.definition["http_headers"]:
                value = safe_eval.safe_eval(h["value"], context)
                res.update({
                        h["name"]: value
                    })
        except:
            _logger.warning("Exception evaluating header value.", exc_info=True)

        return res
